### PR TITLE
"Kickout" and early stopping for translation model training

### PR DESCRIPTION
**Pull request by mjdenkowski**
This adds an option to "kick out" (exclude) training instances each epoch according to a user-specified file of rates.  At the start of each epoch, each sentence is considered for exclusion.  A keep rate of 1.0 guarantees inclusion while a rate of 0.2 would only have a 20% chance of inclusion.  After kickout is applied, minibatches for the epoch are created and shuffled.  Since exclusion is random subject to the rates, different epochs will have different subsets of the data.

One obvious use of kickout is to conduct oversampling as in Google's recent paper.  Doing this by including many instances of the minority data would lead to all instances landing in the same minibatch, which is probably not desirable.  Doing this by trimming down the majority data in advance would make hard exclusion decisions, which is also probably not desirable.  Using kickout to have shorter epochs that always include all minority data and a randomly sampled subset of majority data gets around these issues.

I also added a very basic option for early stopping, which people may or may not find useful, but it was only a few lines of code.

### CODE REVIEW

**Comment by neubig** [comment]
Nice idea, thanks! I've merged it for now. We eventually should really clean up the training code to make the function calls more sane (by creating a "training parameters" struct or something), but maybe that's something for another day.
